lib/forth_evaluator/parser.ex

```diff
@@ -0,0 +1,160 @@
+  # If successful, returns a tuple with the matching tokens and the remaining program
+  # that was not able to parse (an empty list if none).
+  @spec program_parser(words :: [word]) :: {[token], [word]} | :error
+  defp program_parser(words) do
+    [&stack_ops_parser/1, &definition_ops_parser/1, &evaluation_ops_parser/1]
+    |> Combinators.any()
+    |> Combinators.repeat()
+    |> apply([words])
+  end
+
+  # Parser that parses a program given as a list of words.
+  # This parser matches only numbers and, if successful, returns a tuple 
+  # with a "push" `stack_op` for the matching number and the remaining 
+  # program that was not able to parse (an empty list if none).
+  @spec number_parser(words :: [word]) :: {stack_op, [word]} | :error
+  defp number_parser(words)
+
+  defp number_parser([]), do: :error
+
+  defp number_parser([head | tail]) do
+    case Integer.parse(head) do
+      {number, ""} -> {{:stack_op, :push, [number]}, tail}
+      _ -> :error
+    end
+  end
+
+  # Parser that parses a program given as a list of words.
+  # This parser matches all predefined operations and, if successful, returns 
+  # a tuple with the corresponding `stack_op` and the remaining 
+  # program that was not able to parse (an empty list if none).
+  defp operation_parser(words)
+
+  defp operation_parser([]), do: :error
+
+  defp operation_parser([head | tail]) do
+    case String.upcase(head) do
+      "+" -> {{:stack_op, :add, []}, tail}
+      "-" -> {{:stack_op, :substract, []}, tail}
+      "*" -> {{:stack_op, :multiply, []}, tail}
+      "/" -> {{:stack_op, :divide, []}, tail}
+      "DUP" -> {{:stack_op, :duplicate, []}, tail}
+      "DROP" -> {{:stack_op, :drop, []}, tail}
+      "SWAP" -> {{:stack_op, :swap, []}, tail}
+      "OVER" -> {{:stack_op, :over, []}, tail}
+      "." -> {{:stack_op, :pop, []}, tail}
+      _ -> :error
+    end
+  end
+
+  # Parser that parses a program given as a list of words.
+  # This parser matches any operations (numbers and predefined operations)
+  # and, if successful, returns a tuple with the corresponding `stack_op` and the remaining 
+  # program that was not able to parse (an empty list if none).
+  defp stack_ops_parser(words) do
+    [&operation_parser/1, &number_parser/1]
+    |> Combinators.any()
+    |> Combinators.repeat_once()
+    |> apply([words])
+  end
+
+  # Parser that parses a program given as a list of words.
+  # This parser matches new word definitions and, if successful, returns 
+  # a tuple with the corresponding `definition_op` and the remaining 
+  # program that was not able to parse (an empty list if none).
+  defp definition_ops_parser(words) do
+    colon_parser = Combinators.consume(&parse_match(":", &1))
+    semicolon_parser = Combinators.consume(&parse_match(";", &1))
+    body_parser = Combinators.any([&stack_ops_parser/1, &evaluation_ops_parser/1])
+
+    case [colon_parser, &name_parser/1, body_parser, semicolon_parser]
```

Comment:
is forth supposed to support calling definitions within definitions? For example something like:

```
    result = ForthEvaluator.Parser.parse_program(": square DUP * ;\n: doublesquare square square ;\n3 doublesquare .")
```

(I hope I haven't messed up the syntax, i'm not too used to forth)

I have a feeling this might not support it because evaluation_ops_parser can only match a single evaluation (note that stack_ops_parser is able to match multiple operands)

Reply:
You are correct, the current implementation here does not support definitions within definitions. This feature was added in #6 which made it easier to update.